### PR TITLE
Fix ds-identify's read_fs_info() to work on FreeBSD

### PR DESCRIPTION
**Pull request by igalic**
previously, ds-identify would mostly just guess that the system in question is `maybe` OpenStack.

This patch makes it correctly recognize NoCloud.

## Proposed Commit Message

>  Fix ds-identify's read_fs_info() to work on FreeBSD
>
> FreeBSD doesn't have blkid, so we want to use geom to list devices and their fstypes and labels.
>
>This PR also adds `jail` to the list of is_container()
>And we now also properly cache geom and blkid output!
>
> Finally, A test is added to verify the new behaviour by correctly identifying NoCloud on FreeBSD.

## Additional Context

This fixes LP 1901174

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html)
 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

File: tools/ds-identify [hunk 1]
```diff
@@ -281,6 +280,49 @@ read_fs_info() {
...
+
+    # shellcheck disable=2086
+    { IFS="$CR"; set -- $out; IFS="$oifs"; }
+
+    for line in "$@"; do
+        set -- $line
+        provider=$1
+        ftype="${provider%/*}"
+        label="${provider#*/}"
+	dev=$3
```
Comment (smoser) [inline tools/ds-identify:302]: white space.

File: tools/ds-identify [hunk 2]
```diff
@@ -281,6 +280,49 @@ read_fs_info() {
     DI_ISO9660_DEVS="${isodevs#,}"
 }
 
+read_geom() {
+    local oifs="$IFS" line="" delim=","
+    local ret=0 out="" labels="" dev="" label="" ftype="" isodevs="" uuids=""
```
Comment (smoser) [inline tools/ds-identify:285]: check if geom is available?
command -v geom

File: tools/ds-identify [hunk 3]
```diff
@@ -281,6 +280,49 @@ read_fs_info() {
...
+        error "failed running [$ret]: geom label status -as"
+        DI_FS_LABELS="$UNAVAILABLE:error"
+        DI_ISO9660_DEVS="$UNAVAILABLE:error"
+        return $ret
+    }
+
+    # shellcheck disable=2086
+    { IFS="$CR"; set -- $out; IFS="$oifs"; }
+
+    for line in "$@"; do
```
Comment (smoser) [inline tools/ds-identify:297]: i'd like to see an example of what these lines look like.

**Review by smoser — COMMENTED** [review]
needs tests in test_dsidentify.

also... maybe we consider trying to render different versions ? freebsd and linux ?

rather than trying linux then trying freebsd.

probably that is not necessary if we're doing it based on uname.

File: tools/ds-identify (hunk 1 above)
Comment (igalic) [reply tools/ds-identify:302]: thanks!

(this was written in an unconfigured nvim on FreeBSD, and then just transferred over…)

File: tools/ds-identify (hunk 2 above)
Comment (igalic) [reply tools/ds-identify:285]: it's in `/sbin/geom`

which means it's in "base".
i don't even see a `WITHOUT_GEOM`: https://www.freebsd.org/cgi/man.cgi?src.conf
so i don't think you can build a modern FreeBSD without geom

File: tools/ds-identify (hunk 3 above)
Comment (igalic) [reply tools/ds-identify:297]:
```
meena@fbsd12-1 /u/h/meena> geom label status -a
          Name Status Components
  gpt/gptboot0 N/A vtbd1p1
     gpt/swap0 N/A vtbd1p2
iso9660/cidata N/A vtbd2
```

and here, in -s for script-friendly output:

```
meena@fbsd12-1 /u/h/meena> geom label status -as
  gpt/gptboot0 N/A vtbd1p1
     gpt/swap0 N/A vtbd1p2
iso9660/cidata N/A vtbd2
```

File: tools/ds-identify [hunk 4]
```diff
@@ -281,6 +280,50 @@ read_fs_info() {
...
+
+        labels="${labels}${label}${delim}"
+    done
+
+    DI_FS_LABELS="${labels%${delim}}"
+    DI_ISO9660_DEVS="${isodevs#,}"
+}
+
+read_fs_info() {
+    cached "${DI_BLKID_OUTPUT}" && return 0
```
Comment (smoser) [inline tools/ds-identify:316]: bah. this needs fixing. we dont actually do anything to set that variable.
the intent of the variable is to say we've already run blkid and cached its output.

this isn't your fault. the intent of DI_BLKID_OUTPUT was to store the result of calling 'blkid', so that we didn't call that external program more than once.  So I guess what I'd like from you:

a.) a comment in read_fs_info stating what variables it sets (DI_FS_LABELS, DI_ISO9660_DEVS, DI_FS_UUIDS)
b.) drop of the DI_BLKID_OUTPUT cache here
c.) add caching of call to 'geom'. the easiest way to do this might be just adding a variable 'DI_GEOM' and put the call to 'geom' in a function, something like:

    geom_label_status_as() {
        # call the bsd program geom '-as', set DI_GEOM_LABEL_STATUS_OUT
        cached "$DI_GEOM_LABEL_STATUS_OUT" && return 0
        local out=""
        out=$(geom -as label status) &&  DI_GEOM_LABEL_STATUS="$out" ||
            DI_GEOM_LABEL_STATUS="$UNAVAILABLE"
        return
     }
         

d.) fixing of the caching of 'blkid' output (i realize i'm asking you to fix a bug that isn't your fault... go ahead and git blame this and see that it is mine. :wink: ), but having this variable/cache that doesn't do anything is just confusing.

File: tools/ds-identify [hunk 5]
```diff
@@ -281,6 +280,50 @@ read_fs_info() {
     DI_ISO9660_DEVS="${isodevs#,}"
 }
 
+read_geom() {
```
Comment (smoser) [inline tools/ds-identify:283]: you can rename this function to 'read_fs_info_freebsd' and the linux one 'read_fs_info_linux'.

File: tools/ds-identify [hunk 6]
```diff
@@ -281,6 +280,50 @@ read_fs_info() {
...
+    local oifs="$IFS" line="" delim=","
+    local ret=0 out="" labels="" dev="" label="" ftype="" isodevs=""
+    out=$(geom label status -as) || {
```
Comment (smoser) [inline tools/ds-identify:286]: after adding the 'geom_label_status_as' function, then you'll just use the variable, so here this would look like:

    geom_label_status_as
    local out="$DI_GEOM_LABEL_STATUS_OUT"
    ...
    [ "$out" ] = "$UNAVAILABLE" && return 0
    
    and the rest of your function here.

**Review by smoser — CHANGES_REQUESTED** [review]

File: tools/ds-identify (hunk 4 above)
Comment (igalic) [reply tools/ds-identify:316]: > i realize i'm asking you to fix a bug that isn't your fault...

i'm perfectly happy to do that, it just means this patch will take as long as all my other patches usually do 😅

**Comment by igalic** [comment]
@smoser any comments on the test?

**Comment by smoser** [comment]
> @smoser any comments on the test?

Sure, 2 comments:
a.) i dont think you actually *added* a test.  (need to add a function name test_)
b.) other than that i thin your changes there look fine.  the 'blkid_out' method that you  mimic'd in "geom_out" was an attempt to be less verbose in tests. it feels like you did the right thing here in translating the data to geom output.

File: tests/unittests/test_ds_identify.py [hunk 7]
```diff
@@ -807,6 +830,19 @@ def _print_run_output(rc, out, err, cfg, files):
             'dev/vdb': 'pretend iso content for cidata\n',
         }
     },
+    'NoCloud-fbsd': {
```
Comment (igalic) [inline tests/unittests/test_ds_identify.py:833]: @smoser i naïvely assumed this would add a test case

File: tools/ds-identify (hunk 6 above)
Comment (igalic) [reply tools/ds-identify:286]: you mean
```sh
[ "$out" = "$UNAVAILABLE" ] && return 0
```
right?

File: tools/ds-identify [hunk 8]
```diff
@@ -319,6 +372,11 @@ detect_virt() {
                 *) virt="$out"
             esac
         }
+        out=$(sysctl -qn security.jail.jailed 2>/dev/null) && {
+            if [ "$out" = "1" ]; then
+                virt="jail"
+            fi
+        }
```
Comment (igalic) [inline tools/ds-identify:379]: I'm not sure if this shouldn't be part of a different changeset

File: tools/ds-identify [hunk 9]
```diff
@@ -331,7 +389,7 @@ read_virt() {
 
 is_container() {
     case "${DI_VIRT}" in
-        container-other|lxc|lxc-libvirt|systemd-nspawn|docker|rkt) return 0;;
+        container-other|lxc|lxc-libvirt|systemd-nspawn|docker|rkt|jail) return 0;;
```
Comment (igalic) [inline tools/ds-identify:392]: but if I'm already touching a code path where this function is called, and adapting it to FreeBSD, then i figured I'd make is_container work on (and for!) FreeBSD

**Review by igalic — COMMENTED** [review]
ℹ️

File: tools/ds-identify (hunk 8 above)
Comment (smoser) [reply tools/ds-identify:379]: i'm fine with it here, but maybe call it bsd-json?

File: tools/ds-identify (hunk 6 above)
Comment (smoser) [reply tools/ds-identify:286]: yes. good lint'ing.

File: tools/ds-identify (hunk 8 above)
Comment (igalic) [reply tools/ds-identify:410]: I'm going by what other tools like `facter` return:

```
root@webirc:~ # facter virtual
jail
root@webirc:~ #
```

on a completely unrelated note: while it's possible to run Linux in a FreeBSD jail, the version is so outdated, that our tools would barely pick run on it, so, :woman_shrugging:

File: tools/ds-identify (hunk 4 above)
Comment (igalic) [reply tools/ds-identify]: i *think* this is done now

File: tools/ds-identify (hunk 6 above)
Comment (igalic) [reply tools/ds-identify]: hmmmm…
for now i'm passing on the return value instead of return 0… not sure that's a sensible thing to do

File: tools/ds-identify [hunk 10]
```diff
@@ -231,8 +232,20 @@ ensure_sane_path() {
     done
 }
 
-read_fs_info() {
-    cached "${DI_BLKID_OUTPUT}" && return 0
+blkid_export() {
+    # call 'blkid -c /dev/null export', set DI_BLKID_EXPORT_OUT
+    cached "$DI_BLKID_EXPORT_OUT" && return 0
+    local out=""
+    local ret=0
```
Comment (smoser) [inline tools/ds-identify:239]: you can put these on the same line.

File: tools/ds-identify [hunk 11]
```diff
@@ -244,20 +257,22 @@ read_fs_info() {
...
-    out=$(blkid -c /dev/null -o export) || {
-        ret=$?
-        error "failed running [$ret]: blkid -c /dev/null -o export"
+    local ret=0 labels="" dev="" label="" ftype="" isodevs="" uuids=""
+
+    blkid_export
+    ret=$?
+    [ "$DI_BLKID_EXPORT_OUT" = "$UNAVAILABLE" ] && {
         DI_FS_LABELS="$UNAVAILABLE:error"
         DI_ISO9660_DEVS="$UNAVAILABLE:error"
```
Comment (smoser) [inline tools/ds-identify:266]: what happened to DI_FS_UUIDS

Comment (igalic) [reply tools/ds-identify:266]: they were never there to begin with.
but i can add them, if you so desire.

File: tools/ds-identify (hunk 9 above)
Comment (smoser) [reply tools/ds-identify:423]: can we make that bsd-jail ?

File: tests/unittests/test_ds_identify.py (hunk 7 above)
Comment (smoser) [reply tests/unittests/test_ds_identify.py:833]: still no test, right?

Comment (igalic) [reply tests/unittests/test_ds_identify.py:833]: gimme a day or two, while i work my way thru https://pythontesting.net/start-here/

**Review by smoser — APPROVED** [review]
I like it. good work.
